src/main/java/com/linecorp/armeria/common/reactivestreams/FunctionProcessor.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2016 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.reactivestreams;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import java.util.function.Function;
+
+import org.reactivestreams.Processor;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+public final class FunctionProcessor<T, R> implements Processor<T, R> {
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<FunctionProcessor, Subscription> subscriptionUpdater =
+            AtomicReferenceFieldUpdater.newUpdater(FunctionProcessor.class, Subscription.class, "subscription");
+
+    @SuppressWarnings({ "rawtypes", "AtomicFieldUpdaterIssues" })
+    private static final AtomicReferenceFieldUpdater<FunctionProcessor, Subscriber> subscriberUpdater =
+            AtomicReferenceFieldUpdater.newUpdater(FunctionProcessor.class, Subscriber.class, "subscriber");
```

Comment:
When I researched this myself, I found this excellent article from the netty team :)
http://normanmaurer.me/blog/2013/10/28/Lesser-known-concurrent-classes-Part-1/